src/Components/Browser.JS/src/Platform/Circuits/RenderQueue.ts

```diff
@@ -0,0 +1,106 @@
+import { renderBatch } from '../../Rendering/Renderer';
```

Comment:
Not sure if I'm missing something critical, but could we avoid the need for having a queue entirely? By the time data reaches the client, rendering is synchronous and has to take near-zero time otherwise the user experience is broken.

Can we avoid buffering things up and tracking states by rendering directly as we receive the data? I know we do need to track the last-applied batch ID so we can skip dupes, but that on its own doesn't require maintaining a queue.

As far as I can tell, the only reason the current code separates "enqueing" from "rendering" is so it can send back `OnRenderCompleted` before is does the "rendering" phase. If that's the only reason I think we can do without the extra abstraction.

Reply:
Batches can arrive out of order when they are sent after a re-connection event as we simply trigger the sending for all the batches. Maybe SignalR serializes all messages sent, but I don't want to rely on that assumption in this code.

https://github.com/aspnet/AspNetCore/pull/8911/files#diff-d33c04a2c673816a9a5c6e959a8e5270R152

> Not sure if I'm missing something critical, but could we avoid the need for having a queue entirely? By the time data reaches the client, rendering is synchronous and has to take near-zero time otherwise the user experience is broken.

Reply:
Also, are we accepting that there will always be a single circuit per browser instance and a single browserRendererId per circuit? I wrote this code without making those assumptions. If we are OK with making that statement, then we can probably simplify this.

Even if we simplify the code, I would like to keep this abstraction in place (as a static class if you want) just so that all the code dealing with the rendering is not mixed up with the rest of Boot.Server.ts which is growing boundlessly and accumulating concerns.

Reply:
We *strongly* rely on SignalR preserving message order :) If this is not guaranteed, we are badly broken in other ways too.

https://github.com/aspnet/AspNetCore/blob/8759aa2fc04a896e2a27fdc057e8f5c55db431d0/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L94

Given this, can we eliminate the extra layer of queuing?

Reply:
> Also, are we accepting that there will always be a single circuit per browser instance and a single browserRendererId per circuit?

We're probably safe with "single browserRenderId per circuit" since there's no API that lets you get outside that scenario, nor do we plan to add one. However there could be multiple circuits per browser if we choose to let people connect to multiple distinct endpoints altogether, which I think is likely.

> Even if we simplify the code, I would like to keep this abstraction in place (as a static class if you want)

That sounds good to me! A file containing a bunch of `export function`s is a nice way to split things up.

Reply:
Removed the storing incoming batches in a queue and simplified this to just process the batches right away.